### PR TITLE
Add BatchNormalization

### PR DESCRIPTION
**Pull request by KSTconsult**

### CODE REVIEW

**Comment by KSTconsult** [comment]
@l-bat @RodionovF @TolyaTalamanov @apertovs

File: src/runtime/cpu/kernels/kernels.cpp [hunk 1]
```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+                                    Vector running_mean, Vector running_var,
+                                    bool isTraining, float alpha,
+                                    ConstVector gamma, ConstVector beta) {
+
+    if (isTraining) {
+        auto input_mean = input.colwise().mean();
+        input_centered = input.rowwise() - input_mean;
+
+        auto input_var = input_centered.cwiseAbs2().colwise().mean();
+        std = (input_var.array() + 0.001).cwiseSqrt();
```
Comment (l-bat) [inline src/runtime/cpu/kernels/kernels.cpp:93]: Please use epsilon instead of 0.001

File: src/runtime/cpu/kernels/kernels.hpp [hunk 2]
```diff
@@ -105,4 +105,37 @@ std::vector<int> MatchTargetTo1dMatrix(ConstVector target, int batch_size, int n
  */
 float CPUNLLLoss(Matrix predictions, ConstVector target);
 
+/*
+ * CPUBatchNormForward
+ * Implements Batch Normalization forward pass
+ * input have size [batch_size, in_features]
```
Comment (l-bat) [inline src/runtime/cpu/kernels/kernels.hpp:111]: We need BatchNorm1D and BatchNorm2D. Please look at Pytorch.

Comment (l-bat) [reply src/runtime/cpu/kernels/kernels.hpp:111]: Lets rename `CPUBatchNormForward` to `CPUBatchNorm1DForward`

File: src/expression/nn.cpp [hunk 3]
```diff
@@ -45,3 +45,43 @@ deepworks::Softmax::Softmax(std::string name)
...
+void deepworks::BatchNorm::init(const Shape& in_shape) {
+
+    // NB: Init gamma
+    deepworks::Tensor gamma(deepworks::Shape{in_shape[1]});
+    deepworks::initializer::constant(gamma, 0.2);
+    m_info.impl().params.emplace_back(std::move(gamma), true);
+
+    // NB: Init beta.
+    deepworks::Tensor beta(deepworks::Shape{in_shape[1]});
+//    deepworks::initializer::zeros(beta);
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:64]: Still needed ?

File: src/runtime/cpu/layers/cpubatchnorm.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                  std::vector<Tensor>& outputs);
+
+
+    deepworks::Tensor m_gamma, m_beta;
+    deepworks::Tensor m_input_centered, m_std;
+    deepworks::Tensor m_running_mean, m_running_var;
+};
+
+} // namespace cpu
+} // namespace deepworks
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpubatchnorm.hpp:29]: Broken line

File: src/expression/nn.cpp [hunk 5]
```diff
@@ -45,3 +45,43 @@ deepworks::Softmax::Softmax(std::string name)
...
+    deepworks::initializer::constant(beta, 0.1);
+    m_info.impl().params.emplace_back(std::move(beta), true);
+
+    deepworks::Tensor input_centered(in_shape);
+    deepworks::initializer::zeros(input_centered);
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:69]: I believe `input_centered`, `std`, `running_mean`, `running_var` - are't `Paramaters` at all, because the aren't trainable, it's just a temporary buffers.

So it can be placed in `CPUBactchnorm1D`

File: src/runtime/cpu/kernels/kernels.cpp [hunk 6]
```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
     float loss = -X_1d(0, slice).array().sum() / static_cast<float>(batch_size);
     return loss;
 }
+
+void deepworks::CPUBatchNorm1DForward(ConstMatrix input, Matrix output,
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:82]: Please separate computing gradients by inputs from parameters, because if layer is frozen it isn't necessary to compute parameters gradients but computing gradient by inputs is mandatory and vice versa if the layer is input no need to compute gradient by inputs, but if isn't frozen need to compute by parameters.

Have a look on PR with `backward`

File: src/runtime/cpu/kernels/kernels.cpp [hunk 7]
```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+                                      Matrix input_centered, Vector std,
+                                      Vector running_mean, Vector running_var,
+                                      bool isTraining, float eps, float alpha,
+                                      ConstVector gamma, ConstVector beta) {
+
+    if (isTraining) {
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:88]: no camel case, please :)

File: src/runtime/cpu/kernels/kernels.cpp [hunk 8]
```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+        running_mean = running_mean * alpha + input_mean * (1 - alpha);
+        running_var = running_var * alpha + input_var * (1 - alpha);
+
+        // FIXME: holy shit, if someone knows how to do it beautifully, tell me
+        output = ((input_centered.array().rowwise() / std.array())
+                 .array().rowwise() * gamma.array()).array().rowwise() + beta.array();
+
+    } else {
+        // FIXME: write in local variable in eval mod normal?
+        auto input_centered = input.rowwise() - running_mean;
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:104]: Does it shadow function parameters ?

**Review by TolyaTalamanov — COMMENTED** [review]
Don't dive deep in `Eigen` implementation, I believe other reviewers do

File: src/expression/nn.cpp (hunk 5 above)
Comment (TolyaTalamanov) [reply src/expression/nn.cpp:69]: @l-bat Correct me

File: src/runtime/cpu/kernels/kernels.cpp [hunk 9]
```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+        auto std = (running_var.array() + eps).cwiseSqrt();
+
+        output = ((input_centered.array().rowwise() / std).array().rowwise() * gamma.array()).array().rowwise() + beta.array();
```
Comment (l-bat) [inline src/runtime/cpu/kernels/kernels.cpp:107]: Does this line depend on "training"?

File: src/runtime/cpu/kernels/kernels.cpp [hunk 10]
```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+
+        auto input_var = input_centered.cwiseAbs2().colwise().mean();
+        std = (input_var.array() + eps).cwiseSqrt();
+
+        running_mean = running_mean * alpha + input_mean * (1 - alpha);
+        running_var = running_var * alpha + input_var * (1 - alpha);
+
+        // FIXME: holy shit, if someone knows how to do it beautifully, tell me
+        output = ((input_centered.array().rowwise() / std.array())
+                 .array().rowwise() * gamma.array()).array().rowwise() + beta.array();
```
Comment (RodionovF) [inline src/runtime/cpu/kernels/kernels.cpp:100]: the same, but less `.array().rowwise()`

```python
(input_centered.array().rowwise() * (gamma.array() / std.array())).array().rowwise() + beta.array();
```

Comment (RodionovF) [reply src/runtime/cpu/kernels/kernels.cpp:100]: imho, it's not necessary to pass `Matrix input_centered, Vector std`. It is not difficult to count them in Backward, and the function's signature will become cleaner. But this is a moment for discussion.

also you can use (gamma.array() / std), if you set `auto` type for `input_centered` and `std`.

Comment (RodionovF) [reply src/runtime/cpu/kernels/kernels.cpp:100]: type when use auto (BinaryOp) auto

Comment (RodionovF) [reply src/runtime/cpu/kernels/kernels.cpp:100]: if pass through the function pass_by

File: src/expression/nn.cpp (hunk 5 above)
Comment (KSTconsult) [reply src/expression/nn.cpp]: Done

**Review by TolyaTalamanov — COMMENTED** [review]
According to the latest master, please provide implementation for `backward` in `CPUBackend` as well

**Comment by l-bat** [comment]
Your branch contains a lot of commits from main. Please rebase on main.

**Comment by l-bat** [comment]
relates to #28
